File: src/Point.js

"use strict";
class Point {
    static width = 10000;
    static xcenter = 5000;
    static ycenter = 5000;

    static from(x, y) { return (y + Point.ycenter) * Point.width + (x + Point.xcenter); }
    static to2D(p) { return [p % Point.width - Point.xcenter, Math.floor(p / Point.width) - Point.ycenter]; }

    static up(p, times = 1) { return p - times * Point.width; }
    static down(p, times = 1) { return p + times * Point.width; }
    static left(p, times = 1) { return p - times; }
    static right(p, times = 1) { return p + times; }
    static upLeft(p, times = 1) { return p - times - times * Point.width; }
    static upRight(p, times = 1) { return p + times - times * Point.width; }
    static downLeft(p, times = 1) { return p - times + times * Point.width; }
    static downRight(p, times = 1) { return p + times + times * Point.width; }

    static cardinals(p, times = 1) { return [Point.up(p, times), Point.down(p, times), Point.left(p, times), Point.right(p, times)]; }
    static diagonals(p, times = 1) { return [Point.upLeft(p, times), Point.upRight(p, times), Point.downLeft(p, times), Point.downRight(p, times)]; }
    static neighborhood(p, times = 1) { return [...Point.cardinals(p, times), ...Point.diagonals(p, times)]; }

    static distance(a, b) { return Math.abs(a % Point.width - b % Point.width) + Math.floor(Math.abs(a - b) / Point.width); }
    static toString(p) { return Point.to2D(p).toString() }
}

module.exports = {
    Point
}

File: src/Point.cs

namespace Powers
{
    public class Point
    {
        public readonly int X;
        public readonly int Y;

        public Point(int x = 0, int y = 0)
        {
            this.X = x;
            this.Y = y;
        }

        public Point(Point point) : this(point.X, point.Y) { }

        public int compareX(Point other)
        {
            return X - other.X;
        }

        public int compareY(Point other)
        {
            return Y - other.Y;
        }

        public int distance(Point other)
        {
            return (int)Math.Max(Math.Abs(X - other.X), MathF.Floor(Math.Abs(Y - other.Y)));
        }

        public Point Up(int times = 1)
        {
            return new Point(X, Y - times);
        }

        public Point Down(int times = 1)
        {
            return new Point(X, Y + times);
        }

        public Point Left(int times = 1)
        {
            return new Point(X - times, Y);
        }

        public Point Right(int times = 1)
        {
            return new Point(X + times, Y);
        }

        public Point UpLeft(int times = 1)
        {
            return new Point(X - times, Y - times);
        }

        public Point UpRight(int times = 1)
        {
            return new Point(X + times, Y - times);
        }

        public Point DownLeft(int times = 1)
        {
            return new Point(X - times, Y + times);
        }

        public Point DownRight(int times = 1)
        {
            return new Point(X + times, Y + times);
        }

        public List<Point> Cardinals(int times = 1)
        {
            return new List<Point> { Up(times), Down(times), Left(times), Right(times) };
        }

        public List<Point> Diagonals(int times = 1)
        {
            return new List<Point> { UpLeft(times), UpRight(times), DownLeft(times), DownRight(times) };
        }

        public List<Point> Neighborhood(int times = 1)
        {
            return new List<Point>{Up(times), Down(times), Left(times), Right(times), UpLeft(times), UpRight(times),
                DownLeft(times), DownRight(times)};
        }

        public override int GetHashCode()
        {
            return (Y << 16) | X;
        }

        public override bool Equals(Object? obj)
        {
            if (obj == null)
            {
                return false;
            }
            Point? point = obj as Point;
            if (point == null)
            {
                return false;
            }
            if (base.Equals(obj))
            {
                return true;
            }
            return X == point.X && Y == point.Y;
        }

        public override string ToString()
        {
            return string.Format("({}, {})", X, Y);
        }
    }
}
